### ISSUE TITLE
fastmri pytorch ResourceExhausted error w torchrun

### ISSUE THREAD

**Issue by priyakasimbeg** [mlcommons/algorithmic-efficiency#182]
Running the fastmri pytorch workload w torch run results in a resource exhausted error. 

## Description

Running the fastmri pytorch workload w torch run results in a resource exhausted error (on the host). Running without torchrun seems to work fine. 

Traceback:
```
I1012 02:23:44.180147 140690949699392 spec.py:287] Evaluating on the training split.
I1012 02:24:16.079694 140690949699392 spec.py:299] Evaluating on the validation split.
I1012 02:24:26.101415 140690949699392 spec.py:314] Evaluating on the test split.
I1012 02:24:35.252639 140690949699392 submission_runner.py:308] 733.32s         863     {'train/ssim': 0.09793876231666465, 'train/loss': 0.042989944809208015, 'validation/ssim': 0.09695557489004995, 'validation/loss': 0.046180981780808227, 'test/ssim': 0.09845683796602489, 'test/loss': 0.046397272820522244}
I1012 02:24:35.261649 140618806310656 logging_writer.py:48] [863] global_step=863, test/loss=0.046397, test/ssim=0.098457, train/loss=0.042990, train/ssim=0.097939, validation/loss=0.046181, validation/ssim=0.096956
I1012 02:25:04.387545 140690949699392 spec.py:287] Evaluating on the training split.
I1012 02:25:32.222669 140690949699392 spec.py:299] Evaluating on the validation split.
2022-10-12 02:36:44.085076: W tensorflow/core/framework/op_kernel.cc:1733] RESOURCE_EXHAUSTED: MemoryError: Unable to allocate 1.82 MiB for an array with shape (640, 372) and data type complex64
Traceback (most recent call last):

  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/tensorflow/python/ops/script_ops.py", line 270, in __call__
    ret = func(*args)

  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/tensorflow/python/autograph/impl/api.py", line 642, in wrapper
    return func(*args, **kwargs)

  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/tensorflow/python/data/ops/dataset_ops.py", line 1030, in generator_py_func
    values = next(generator_state.get_iterator(iterator_id))

  File "/home/kasimbeg/algorithmic-efficiency/algorithmic_efficiency/workloads/fastmri/input_pipeline.py", line 134, in _h5_to_examples
    yield hf['kspace'][i], hf['kspace'][i].shape, hf['reconstruction_esc'][

  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper

  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper

  File "/home/kasimbeg/algorithmic-efficiency/env/lib/python3.7/site-packages/h5py/_hl/dataset.py", line 819, in __getitem__
    arr = numpy.zeros(selection.array_shape, new_dtype, order='C')

numpy.core._exceptions._ArrayMemoryError: Unable to allocate 1.82 MiB for an array with shape (640, 372) and data type complex64


Fatal Python error: Segmentation fault
```

## Steps to Reproduce
On vm kasimbeg-2 :
```
$ source ~/algorithmic-efficiency/env/bin/activate
$ torchrun --redirects 1:0,2:0,3:0,4:0,5:0,6:0,7:0 --standalone --nnodes=1 --nproc_per_node=8 submission_runner.py \
    --framework=pytorch \
    --data_dir=/home/kasimbeg/local_data/fastmri  \
    --experiment_dir=$ROOT_DIR \
    --experiment_name=target_setting \
    --workload=fastmri \
    --submission_path=target_setting_runs/pytorch_nadamw.py \
    --tuning_search_space=target_setting_runs/fastmri/tuning_search_space.json
```

## Source or Possible Fix
Running without torchrun seems to be working fine. I'm not sure what in the torchrun workflow is causing this OOM.

**Comment by priyakasimbeg** [comment]
I pulled from main and reran on 2 of my vms.

Running on kasimbeg-2 gave a CUDA out of memory error.

I ran three times on kasimbeg-1. The first time it got to at least 8 epochs. After stopping and restarting the run (to log it to a text file) I got these errors:
First on GPU 7, 2 and 3 and second on 3, 2, 6, 4:
```
RuntimeError: CUDA out of memory. Tried to allocate 6.05 GiB (GPU 7; 15.78 GiB total capacity; 1.54 GiB already allocated; 6.05 GiB free; 8.28 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
W1019 19:13:49.873734 140551252768576 submission_runner.py:350] error: GPU out of memory during eval during step 1, error : CUDA out of memory. Tried to allocate 6.05 GiB (GPU 7; 15.78 GiB total capacity; 1.54 GiB already allocated; 6.05 GiB free; 8.28 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

I'm wondering if something wasn't cleaned up okay but nvidia-smi looks ok:

+-----------------------------------------------------------------------------+
| NVIDIA-SMI 510.47.03    Driver Version: 510.47.03    CUDA Version: 11.6     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla V100-SXM2...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   40C    P0    57W / 300W |      0MiB / 16384MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-SXM2...  Off  | 00000000:00:05.0 Off |                    0 |
| N/A   38C    P0    53W / 300W |      0MiB / 16384MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  Tesla V100-SXM2...  Off  | 00000000:00:06.0 Off |                    0 |
| N/A   40C    P0    55W / 300W |      0MiB / 16384MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  Tesla V100-SXM2...  Off  | 00000000:00:07.0 Off |                    0 |
| N/A   38C    P0    57W / 300W |      0MiB / 16384MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   4  Tesla V100-SXM2...  Off  | 00000000:00:08.0 Off |                    0 |
| N/A   39C    P0    59W / 300W |      0MiB / 16384MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   5  Tesla V100-SXM2...  Off  | 00000000:00:09.0 Off |                    0 |
| N/A   40C    P0    60W / 300W |      0MiB / 16384MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   6  Tesla V100-SXM2...  Off  | 00000000:00:0A.0 Off |                    0 |
| N/A   42C    P0    58W / 300W |      0MiB / 16384MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   7  Tesla V100-SXM2...  Off  | 00000000:00:0B.0 Off |                    0 |
| N/A   40C    P0    60W / 300W |      0MiB / 16384MiB |      3%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |

**Comment by znado** [comment]
With `torchrun`, the shapes of [`logits, targets, volume_max`](https://github.com/mlcommons/algorithmic-efficiency/blob/main/algorithmic_efficiency/workloads/fastmri/ssim.py#L42) are `(32, 320, 320), (32, 320, 320), (32,)` and the SSIM never gets above 0.11, but running with just `python` they are `(256, 320, 320), (256, 320, 320), (256,)` and the SSIM quickly climbs.

**Comment by runame** [comment]
The shapes with DDP (`torchrun`) are exactly as expected, as each process is computing the SSIM for $\frac{1}{\text{NGPUS}}$ of the batch size, which is `256`. Unless I'm missing something it should theoretically be completely fine to compute the SSIM like this, since it is calculated per-example anyway.

One more thing I have noticed: the train/validation loss _also_ seems to be too low, it is around 0.11 with DDP and only around 0.26 with DP or Jax after about the same number of steps. However, the factor is different, so it does not seem to be a wrong denominator.

**Comment by runame** [comment]
Seems like porting the SSIM calculation to PyTorch was sufficient to fix the issue, see PR #214. On GCP it ran successful and finished with a score of 6599.214844.
